PKG-4701 initial v12.4

danpetry opened the pull request:
**Destination channel:** defaults

### Links

- [PKG-4701](https://anaconda.atlassian.net/browse/PKG-4701)

### Explanation of changes:

- Part of CUDA 12.4 build, following [conda-forge's pattern](https://github.com/conda-forge/staged-recipes/issues/21382) which was developed in conjunction with Nvidia. I don't see a need to stage these then release them all at once. They won't cause a problem released one at a time.
- Please check the main branch too, this has just been forked
- Changes made to adjust feedstocks for differences between conda-forge and defaults. Please see commit messages for description of changes.

[PKG-4701]: https://anaconda.atlassian.net/browse/PKG-4701?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

danpetry committed:
Skip s390x, CUDA doesn't support it

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_d9fpn3jezd/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:0: no_global_test: Global tests are ignored in multi-output recipes.
- clone/recipe/meta.yaml:106: license_file_overspecified: Using license_file and license_url is overspecified.
- clone/recipe/meta.yaml:46: host_section_needs_exact_pinnings: output "cuda-opencl": Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:46: host_section_needs_exact_pinnings: output "cuda-opencl": Linked libraries host should have exact version pinnings.

===== ERRORS =====
- clone/recipe/meta.yaml:34: outputs_not_unique: output "cuda-opencl": Output name is not unique
- clone/recipe/meta.yaml:106: missing_dev_url: The recipe is missing a dev_url
- clone/recipe/meta.yaml:106: missing_license_family: The recipe is missing the `about/license_family` key.
- clone/recipe/meta.yaml:1: missing_tests: output "cuda-opencl": No tests were found.
===== Final Report: =====
4 Errors and 4 Warnings were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_338y_wmafw/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:106: license_file_overspecified: Using license_file and license_url is overspecified.
- clone/recipe/meta.yaml:0: no_global_test: Global tests are ignored in multi-output recipes.

===== ERRORS =====
- clone/recipe/meta.yaml:34: outputs_not_unique: output "cuda-opencl": Output name is not unique
- clone/recipe/meta.yaml:106: missing_license_family: The recipe is missing the `about/license_family` key.
- clone/recipe/meta.yaml:1: missing_tests: output "cuda-opencl": No tests were found.
- clone/recipe/meta.yaml:106: missing_dev_url: The recipe is missing a dev_url
===== Final Report: =====
4 Errors and 2 Warnings were found

</details>

danpetry committed:
Correct upstream's bld.bat

- the "if" from "if not exist" wasn't present, resulting in a script failure
- the files in the include\ directory weren't being moved because windows
  doesn't move recursively by default

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_82ciqtae1e/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:106: license_file_overspecified: Using license_file and license_url is overspecified.
- clone/recipe/meta.yaml:0: no_global_test: Global tests are ignored in multi-output recipes.

===== ERRORS =====
- clone/recipe/meta.yaml:106: missing_dev_url: The recipe is missing a dev_url
- clone/recipe/meta.yaml:1: missing_tests: output "cuda-opencl": No tests were found.
- clone/recipe/meta.yaml:34: outputs_not_unique: output "cuda-opencl": Output name is not unique
- clone/recipe/meta.yaml:106: missing_license_family: The recipe is missing the `about/license_family` key.
===== Final Report: =====
4 Errors and 2 Warnings were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_8ez3aljxry/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:106: license_file_overspecified: Using license_file and license_url is overspecified.
- clone/recipe/meta.yaml:0: no_global_test: Global tests are ignored in multi-output recipes.

===== ERRORS =====
- clone/recipe/meta.yaml:106: missing_dev_url: The recipe is missing a dev_url
- clone/recipe/meta.yaml:106: missing_license_family: The recipe is missing the `about/license_family` key.
- clone/recipe/meta.yaml:34: outputs_not_unique: output "cuda-opencl": Output name is not unique
- clone/recipe/meta.yaml:1: missing_tests: output "cuda-opencl": No tests were found.
===== Final Report: =====
4 Errors and 2 Warnings were found

</details>

cbouss reviewed (commented):
Will it work without an ocl loader?
I see this one too: https://github.com/conda-forge/ocl-icd-system-feedstock/tree/main
Which seems to indicate having a system ocl would work out of the box.
Is that package (cuda-opencl) needed for other tasks?

danpetry commented:
The loader's a broker between different OpenCL implementations and your application. It's needed if there are multiple opencl implementations on the system. Which is a common occurrence AFAIU. It's possible to build applications that link directly to the OpenCL implementation but then it will only work with that one which seems to defeat the purpose (it's supposed to allow portability between different hardware architectures). Having said that, cuda-opencl is called by the ICD loader, not the other way around. So I'm not sure if it's accurate to express the ICD loader as a dependency of cuda-opencl, rather should be up to the user to manage this aspect. Even though in practical scenarios it's needed.

danpetry commented:
This package isn't used anywhere else apart from being grouped in the `cuda-libraries` metapackage.

I can see why conda-forge added the loader. I'm open as to whether we should. It doesn't look too difficult.

JeanChristopheMorinPerso commented:
@danpetry DO you have a link to some documentation that could help us understand what's at play? I'm not verry familiar with all of this (read I have zero knowledge).

danpetry commented:
https://www.khronos.org/news/permalink/opencl-installable-client-driver-icd-loader

danpetry commented:
I'll address the dependencies and come back to this discussion if it's looking like a headache for some reason

skupr-anaconda commented on recipe/meta.yaml:
```diff
@@ -20,7 +20,7 @@ source:
 build:
   number: 0
-  skip: true  # [osx or aarch64 or ppc64le]
+  skip: true  # [osx or (linux and s390x) or aarch64 or ppc64le]
```
```suggestion
  skip: true  # [osx or (linux and (s390x or aarch64 or ppc64le))]
```

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_7de4zy8whg/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:46: host_section_needs_exact_pinnings: output "cuda-opencl": Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:0: no_global_test: Global tests are ignored in multi-output recipes.
- clone/recipe/meta.yaml:46: host_section_needs_exact_pinnings: output "cuda-opencl": Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:106: license_file_overspecified: Using license_file and license_url is overspecified.

===== ERRORS =====
- clone/recipe/meta.yaml:34: outputs_not_unique: output "cuda-opencl": Output name is not unique
- clone/recipe/meta.yaml:1: missing_tests: output "cuda-opencl": No tests were found.
- clone/recipe/meta.yaml:106: missing_license_family: The recipe is missing the `about/license_family` key.
- clone/recipe/meta.yaml:106: missing_dev_url: The recipe is missing a dev_url
===== Final Report: =====
4 Errors and 4 Warnings were found

</details>

danpetry committed:
Make platform selector more readable

Co-authored-by: skupr-anaconda <79282962+skupr-anaconda@users.noreply.github.com>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_d3y63993_4/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:106: license_file_overspecified: Using license_file and license_url is overspecified.
- clone/recipe/meta.yaml:46: host_section_needs_exact_pinnings: output "cuda-opencl": Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:46: host_section_needs_exact_pinnings: output "cuda-opencl": Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:0: no_global_test: Global tests are ignored in multi-output recipes.

===== ERRORS =====
- clone/recipe/meta.yaml:1: missing_tests: output "cuda-opencl": No tests were found.
- clone/recipe/meta.yaml:106: missing_dev_url: The recipe is missing a dev_url
- clone/recipe/meta.yaml:106: missing_license_family: The recipe is missing the `about/license_family` key.
- clone/recipe/meta.yaml:34: outputs_not_unique: output "cuda-opencl": Output name is not unique
===== Final Report: =====
4 Errors and 4 Warnings were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(124): `conda run conda-lint /tmp/abs_5bn5j_mkze/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:0: no_global_test: Global tests are ignored in multi-output recipes.
- clone/recipe/meta.yaml:106: license_file_overspecified: Using license_file and license_url is overspecified.
- clone/recipe/meta.yaml:46: host_section_needs_exact_pinnings: output "cuda-opencl": Linked libraries host should have exact version pinnings.
- clone/recipe/meta.yaml:46: host_section_needs_exact_pinnings: output "cuda-opencl": Linked libraries host should have exact version pinnings.

===== ERRORS =====
- clone/recipe/meta.yaml:106: missing_dev_url: The recipe is missing a dev_url
- clone/recipe/meta.yaml:106: missing_license_family: The recipe is missing the `about/license_family` key.
- clone/recipe/meta.yaml:34: outputs_not_unique: output "cuda-opencl": Output name is not unique
- clone/recipe/meta.yaml:1: missing_tests: output "cuda-opencl": No tests were found.
===== Final Report: =====
4 Errors and 4 Warnings were found

</details>
